aws/resource_aws_ec2_capacity_reservation_test.go

```diff
@@ -0,0 +1,222 @@
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckEc2CapacityReservationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig_endDate,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEc2CapacityReservationExists("aws_ec2_capacity_reservation.default", &cr),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "end_date", "2019-10-31T07:39:57Z"),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "end_date_type", "limited"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSEc2CapacityReservation_tags(t *testing.T) {
+	var cr ec2.CapacityReservation
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckEc2CapacityReservationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccEc2CapacityReservationConfig_tags,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckEc2CapacityReservationExists("aws_ec2_capacity_reservation.default", &cr),
+					resource.TestCheckResourceAttr("aws_ec2_capacity_reservation.default", "instance_type", "t2.micro"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckEc2CapacityReservationExists(resourceName string, cr *ec2.CapacityReservation) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).ec2conn
+		resp, err := conn.DescribeCapacityReservations(&ec2.DescribeCapacityReservationsInput{
+			CapacityReservationIds: []*string{aws.String(rs.Primary.ID)},
+		})
+
+		if err != nil {
+			return fmt.Errorf("DescribeCapacityReservations error: %v", err)
+		}
+
+		if len(resp.CapacityReservations) > 0 {
+			*cr = *resp.CapacityReservations[0]
+			return nil
+		}
+
+		return fmt.Errorf("Capacity Reservation not found")
+	}
+}
+
+func testAccCheckEc2CapacityReservationDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).ec2conn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_ec2_capacity_reservation" {
+			continue
+		}
+
+		// Try to find the resource
+		resp, err := conn.DescribeCapacityReservations(&ec2.DescribeCapacityReservationsInput{
+			CapacityReservationIds: []*string{aws.String(rs.Primary.ID)},
+		})
+		if err == nil {
+			for _, r := range resp.CapacityReservations {
+				if *r.State != "cancelled" {
+					return fmt.Errorf("Found uncancelled Capacity Reservation: %s", r)
+				}
+			}
+		}
+
+		return err
+	}
+
+	return nil
+
+}
+
+const testAccEc2CapacityReservationConfig = `
+resource "aws_ec2_capacity_reservation" "default" {
+  instance_type     = "t2.micro"
+  instance_platform = "Linux/UNIX"
+  availability_zone = "us-west-2a"
```

Comment:
Completely optional, but we can allow these configurations to run in any region by utilizing the `aws_availability_zones` data source, e.g.

```hcl
data "aws_availability_zones" "available" {}

availability_zone = "${data.aws_availability_zones.available.names[0]}"
```